### PR TITLE
PUB-1613 System Admin User management dashboard

### PR DESCRIPTION
**Pull request by joshblackmoor**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/PUB-1613

### Change description ###

- Added new endpoint to get all users except third party and filter by all attributes of a user
- Added delete endpoint of a user
- Added update endpoint to update a users role

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 1]
```diff
@@ -383,6 +397,103 @@ public String updateAccount(UserProvenances userProvenance, String provenanceUse
...
+            "User with supplied user id: %s could not be found", userId)));
+
+        deleteAccount(userToDelete.getEmail(), UserProvenances.PI_AAD.equals(userToDelete.getUserProvenance()));
+    }
+
+    /**
+     * Process updating a role for an account.
+     *
+     * @param userId The ID of the user to update.
+     */
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:477]: Just something really minor. Need the second param and the return in the header of this method

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountController.java [hunk 2]
```diff
@@ -241,4 +246,57 @@ public ResponseEntity<Void> deleteExpiredIdamAccounts() {
...
     }
+
+    @ApiResponses({
+        @ApiResponse(responseCode = OK_CODE, description = "Page data with a list of Pi User objects"),
+        @ApiResponse(responseCode = AUTH_ERROR_CODE, description = NOT_AUTHORIZED_MESSAGE)
+    })
+    @Operation(summary = "Get all accounts except third party in a page with filtering")
+    @GetMapping("/all")
+    public ResponseEntity<Page<PiUser>> getAllAccountsExceptThirdParty(
+        @RequestParam(name = "pageNumber") int pageNumber,
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountController.java:257]: Might be worth adding some defaults to number / size just for completeness

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java [hunk 3]
```diff
@@ -685,4 +691,180 @@ void testUpdateAccountUnexpectedDateTimeFormat() {
...
+        List<Roles> emptyRoleList = new ArrayList<>();
+        Page<PiUser> page = new PageImpl<>(List.of(user), pageable, List.of(user).size());
+
+        when(userRepository.findByUserIdPageable(user.getUserId().toString(), pageable)).thenReturn(page);
+
+        Page<PiUser> response = accountService.findAllAccountsExceptThirdParty(pageable, "", "",
+            emptyUserProvenancesList, emptyRoleList, user.getUserId().toString());
+
+        assertEquals(page, response, "Returned page did not match expected");
+    }
```
Comment (KianKwa) [inline src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java:869]: As we return straight from this method after call to find user by ID. Could also verify the second call to the user repository never happen using:

verify(userRepository, never()).findAllByEmailLikeAndUserProvenanceInAndRolesInAndProvenanceUserIdLike(any(), any(), any(), any(), any());

Comment (joshblackmoor) [reply src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java:879]: Done :) Have also updated the other tests with 

verify(userRepository, never()).findByUserIdPageable(any(), any());

When we are not searching by the user id to ensure its not called.

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountController.java (hunk 2 above)
Comment (joshblackmoor) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountController.java]: Done :) has to be a string as thats all defaultValue supports but will be converted to int.

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java (hunk 1 above)
Comment (joshblackmoor) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:479]: Done :)

**Review by KianKwa — APPROVED** [review]

**Review by ChrisS1512 — APPROVED** [review]

**Review by joshblackmoor — COMMENTED** [review]
Tests look good @ashwini-mv good job!

**Review by ChrisS1512 — APPROVED** [review]

**Review by junaidiqbalmoj — APPROVED** [review]
nice work

**Comment by ashwini-mv** [comment]
> Tests look good @ashwini-mv good job!

**Comment by ashwini-mv** [comment]
> > Tests look good @ashwini-mv good job!

**Pull request merged**